### PR TITLE
refactor:return error type for FindCluster in cluster placement strategy interface

### PR DESCRIPTION
**Pull request by JameelB**
## Description
Changes the following methods to return an `error` type instead of the KAS Fleet Manager `ServiceError`. This is based on what was agreed upon in the Control Plane team's technical discussion. Functions should not return `ServiceError` unless there is a strong need for it which wasn't the case for these functions.

Cluster Service
- FindAllClusters: will now return database error as is. This is only called in the cluster placement strategy methods which now returns an `error` type.
- FindCluster: will now return database error as is. This is called in the following places:
   - Cluster placement strategy: This now returns an `error` type.
   - Cluster manager: in the reconcileDeprovisioningCluster, this method already returns an `error` type.
- FindKafkaInstanceCount: will return database error or GetKafkaInstanceSize error as is. This is called in the following places:
   - Cluster placement strategy: This now returns an `error` type.
   - Cluster manager: in the reconcileClusterWithManualConfig and setKafkaPerClusterCountMetrics, both of these methods already returns an `error` type.
- FindStreamingUnitCountByClusterAndInstanceType: Returns database and GetKafkaInstanceSize errors as is. This method already returns an `error` type. This is used for the cluster placement strategy and in the Kafka manager (setClusterStatusCapacityMetrics) which already returns an `error` type.

Cluster Placement Strategy
- FindCluster: interface was changed to return an `error` instead of a `ServiceError`. There wasn't any strong need for it to return a `ServiceError` as the errors it returns will mostly be caused by database errors or configuration error.

Follows up from https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1167#discussion_r925801278

## Verification Steps
- [ ] CI Checks passing

## Checklist (Definition of Done)

- [x] Unit and integration tests added that prove the fix is effective or the feature works (tested against emulated and non-emulated OCM environment)
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer
- [ ] All PR comments are resolved either by addressing them or creating follow up tasks
- [ ] ~~All acceptance criteria specified in JIRA have been completed~~
- [ ] ~~Documentation added for the feature~~
- [ ] ~~Required metrics/dashboards/alerts have been added (or PR created).~~
- [ ] ~~Required Standard Operating Procedure (SOP) is added.~~
- [ ] ~~JIRA has been created for changes required on the client side~~

### CODE REVIEW

**Comment by JameelB** [comment]
please have a look when you get a chance @ziccardi @machi1990 @miguelsorianod @VanillaSpoon

File: internal/kafka/internal/services/cluster_placement_strategy.go [hunk 1]
```diff
@@ -74,7 +74,7 @@ func (f *FirstSchedulableWithinLimit) FindCluster(kafka *dbapi.KafkaRequest) (*a
 	//#1
 	clusterObj, err := f.ClusterService.FindAllClusters(criteria)
 	if err != nil {
-		return nil, err
+		return nil, errors.NewWithCause(errors.ErrorGeneral, err, "failed to find all clusters with criteria: %v", criteria)
```
Comment (VanillaSpoon) [inline internal/kafka/internal/services/cluster_placement_strategy.go]: Hey Jameel, there are a few errors changed back such as:
```diff
-		return nil, err
+		return nil, errors.NewWithCause(errors.ErrorGeneral, err, "failed to find all clusters with criteria: %v", criteria)
```

and then back [in another commit](https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1201/commits/3a6a8673a46df5fb6275b2ca8cf6bfed708b6d1e#diff-5ffcdb3a533058503533bac73799b825a292e9cc5fabc494e6ec10d4d26753f6L77-R76) to:
```diff
-		return nil, errors.NewWithCause(errors.ErrorGeneral, err, "failed to find all clusters with criteria: %v", criteria)
+		return nil, err
```

I was just wondering what the reasoning for this was?

File: internal/kafka/internal/services/cluster_placement_strategy.go [hunk 2]
```diff
@@ -68,7 +67,7 @@ func (f *FirstSchedulableWithinLimit) FindCluster(kafka *dbapi.KafkaRequest) (*a
 
 	kafkaInstanceSize, e := f.KafkaConfig.GetKafkaInstanceSize(kafka.InstanceType, kafka.SizeId)
 	if e != nil {
-		return nil, errors.NewWithCause(errors.ErrorInstancePlanNotSupported, e, "failed to find cluster with criteria '%v'", criteria)
+		return nil, e
```
Comment (ziccardi) [inline internal/kafka/internal/services/cluster_placement_strategy.go:70]: Why this change?

File: internal/kafka/internal/services/cluster_placement_strategy.go [hunk 3]
```diff
@@ -121,9 +120,9 @@ func searchClusterObjInArray(clusters []*api.Cluster, clusterId string) *api.Clu
 }
 
 // findClusterKafkaInstanceCount searches DB for the number of Kafka instance associated with each OSD Clusters
-func (f *FirstSchedulableWithinLimit) findClusterKafkaInstanceCount(clusterIDs []string) (map[string]int, *errors.ServiceError) {
+func (f *FirstSchedulableWithinLimit) findClusterKafkaInstanceCount(clusterIDs []string) (map[string]int, error) {
 	if instanceLst, err := f.ClusterService.FindKafkaInstanceCount(clusterIDs); err != nil {
-		return nil, errors.NewWithCause(err.Code, err, "failed to find kafka instance count for clusters '%v'", clusterIDs)
+		return nil, err
```
Comment (ziccardi) [inline internal/kafka/internal/services/cluster_placement_strategy.go:125]: Why this change?

File: internal/kafka/internal/services/cluster_placement_strategy.go [hunk 4]
```diff
@@ -150,29 +149,31 @@ func (f *FirstReadyWithCapacity) FindCluster(kafka *dbapi.KafkaRequest) (*api.Cl
 	}
 
 	clusters, findAllClusterErr := f.ClusterService.FindAllClusters(criteria)
-	if findAllClusterErr != nil || len(clusters) == 0 {
+	if findAllClusterErr != nil {
```
Comment (ziccardi) [inline internal/kafka/internal/services/cluster_placement_strategy.go:152]: Why did you remove this again?

Comment (JameelB) [reply internal/kafka/internal/services/cluster_placement_strategy.go:152]: I had to remove this due to a previous commit when I changed the return type of `FindAllClusters` but the cluster placement strategy still returns a `ServiceError` so i needed to cast it, which caused tests to fails. 

I'm supposed to re-add this, good spot!

File: internal/kafka/internal/services/cluster_placement_strategy.go (hunk 2 above)
Comment (JameelB) [reply internal/kafka/internal/services/cluster_placement_strategy.go:70]: These changes were due to the change of returned error type in the cluster placement strategy. All of the internal calls now returns an `error` type so we no longer need to create a `ServiceError` out of them. 

We can wrap these errors with the original message, but note that most of these will be database failures.

File: internal/kafka/internal/services/cluster_placement_strategy.go (hunk 1 above)
Comment (JameelB) [reply internal/kafka/internal/services/cluster_placement_strategy.go]: This is to ensure that each commit to refactor each method is isolated. I didn't want to change the return type of a method that I wasn't refactoring. In a later commit, I then refactored the method that calls it to use `error` type instead of a `serviceError` so I no longer had to call `errors.NewWithCause`. 

Isolating each change to a single commit ensures that if we decide to remove one change, we can drop that commit without affecting others that much.

Comment (VanillaSpoon) [reply internal/kafka/internal/services/cluster_placement_strategy.go]: Thanks for the explanation, Very nice tip :)

File: internal/kafka/internal/services/cluster_placement_strategy.go (hunk 2 above)
Comment (machi1990) [reply internal/kafka/internal/services/cluster_placement_strategy.go]: A follow up on @ziccardi question.

> All of the internal calls now returns an error type so we no longer need to create a ServiceError out of them.

the `*errors.ServiceError` implements the `error` interface, I think it'd have been fine to only leave the returned service error.

 > but note that most of these will be database failures
 
 Just to confirm that the database failures are not plainly exposed to the end users, correct? to avoid leaking out db credentials or something ;-)

File: internal/kafka/internal/services/cluster_placement_strategy.go [hunk 5]
```diff
@@ -121,9 +121,9 @@ func searchClusterObjInArray(clusters []*api.Cluster, clusterId string) *api.Clu
 }
 
 // findClusterKafkaInstanceCount searches DB for the number of Kafka instance associated with each OSD Clusters
-func (f *FirstSchedulableWithinLimit) findClusterKafkaInstanceCount(clusterIDs []string) (map[string]int, *errors.ServiceError) {
+func (f *FirstSchedulableWithinLimit) findClusterKafkaInstanceCount(clusterIDs []string) (map[string]int, error) {
 	if instanceLst, err := f.ClusterService.FindKafkaInstanceCount(clusterIDs); err != nil {
-		return nil, errors.NewWithCause(err.Code, err, "failed to find kafka instance count for clusters '%v'", clusterIDs)
+		return nil, err
```
Comment (JameelB) [inline internal/kafka/internal/services/cluster_placement_strategy.go:126]: no need to wrap this one, it should be wrapped by the caller since this is the only place where an error is returned in this method.

File: internal/kafka/internal/services/clusters.go [hunk 6]
```diff
@@ -285,7 +284,7 @@ func (c clusterService) FindCluster(criteria FindClusterCriteria) (*api.Cluster,
 		if errors.Is(err, gorm.ErrRecordNotFound) {
 			return nil, nil
 		}
-		return nil, apiErrors.NewWithCause(apiErrors.ErrorGeneral, err, "failed to find cluster with criteria")
+		return nil, err
```
Comment (JameelB) [inline internal/kafka/internal/services/clusters.go:287]: can return the database error as is since it's the only place that returns an error in this method. The caller of this method can wrap this error instead.

File: internal/kafka/internal/services/clusters.go [hunk 7]
```diff
@@ -415,7 +414,7 @@ func (c clusterService) FindKafkaInstanceCount(clusterIDs []string) ([]ResKafkaI
 	for _, k := range kafkas {
 		kafkaInstanceSize, e := c.kafkaConfig.GetKafkaInstanceSize(k.InstanceType, k.SizeId)
 		if e != nil {
-			return nil, apiErrors.NewWithCause(apiErrors.ErrorInstancePlanNotSupported, e, "failed to query kafkas")
+			return nil, e
```
Comment (JameelB) [inline internal/kafka/internal/services/clusters.go:417]: Didn't see the need to wrap this error as this method returns an already descriptive error message. It returns one of the following error message:
- `Unable to find kafka instance type for '<instance-type>'`
- `Kafka instance size id: '<size-id>' not found for '<region>' instance type`.

The caller of this method can wrap the error returned by this instead.

File: internal/kafka/internal/services/cluster_placement_strategy.go (hunk 2 above)
Comment (JameelB) [reply internal/kafka/internal/services/cluster_placement_strategy.go]: > the *errors.ServiceError implements the error interface, I think it'd have been fine to only leave the returned service error and only change the method signature.

I assumed that we shouldn't mix that two even if we can, maybe I misunderstood that though :) 

> Just to confirm that the database failures are not plainly exposed to the end users, correct? to avoid leaking out db credentials or something ;-)

I think we were wrapping them originally in the Service Errors but only logged them, they're not exposed to users. I believe we return a general error message in the handlers.

File: internal/kafka/internal/services/clusters.go [hunk 8]
```diff
@@ -462,7 +461,7 @@ func (c clusterService) FindAllClusters(criteria FindClusterCriteria) ([]*api.Cl
 		if errors.Is(err, gorm.ErrRecordNotFound) {
 			return nil, nil
 		}
-		return nil, apiErrors.NewWithCause(apiErrors.ErrorGeneral, err, "failed to find all clusters with criteria: %v", clusterDetails)
+		return nil, err
```
Comment (JameelB) [inline internal/kafka/internal/services/clusters.go:464]: same as https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1201#discussion_r927803547

File: internal/kafka/internal/services/cluster_placement_strategy.go (hunk 2 above)
Comment (machi1990) [reply internal/kafka/internal/services/cluster_placement_strategy.go]: > I assumed that we shouldn't mix that two even if we can, maybe I misunderstood that though :)

I was thinking only the method signature may be impacted and we as well as return the service error which is just an implementation of the `error` interface, without any issue. I think we can safely return it and only change the error type from the method signature.

PS: no need to revert the changes here due to explanations you've given above.

Comment (machi1990) [reply internal/kafka/internal/services/cluster_placement_strategy.go]: > I think we were wrapping them originally in the Service Errors but only logged them, they're not exposed to users. I believe we return a general error message in the handlers.

Thanks.

**Review by machi1990 — APPROVED** [review]
Thanks for the followup @JameelB 

I'll let others finish up their reviews and approve as well.

File: internal/kafka/internal/services/cluster_placement_strategy.go (hunk 2 above)
Comment (ziccardi) [reply internal/kafka/internal/services/cluster_placement_strategy.go]: > I assumed that we shouldn't mix that two even if we can, maybe I misunderstood that though :)

What is important is that the returned value is an `error`, which is just an interface.
The returned concrete object can be anything you want as long as it implements `error`.

**Review by ziccardi — APPROVED** [review]
lgtm. Thanks, @JameelB !
